contracts/cw20-merkle-airdrop/src/contract.rs

```diff
@@ -0,0 +1,2158 @@
+        None => return Err(ContractError::CreditsAddress {}),
+    };
+
+    // Update claim index to the current stage
+    CLAIM.save(deps.storage, (proof_addr, stage), &true)?;
+
+    // Update total claimed to reflect
+    let mut claimed_amount = STAGE_AMOUNT_CLAIMED.load(deps.storage, stage)?;
+    claimed_amount += amount;
+    STAGE_AMOUNT_CLAIMED.save(deps.storage, stage, &claimed_amount)?;
+
+    // we stop all airdrops at the date of expiration, and we use the very same date
+    // as a start timestamp for vesting. if expiration is not set, vesting will not work.
+    let vesting_start_time =
+        match expiration {
+            Expiration::AtHeight(_) => {
+                return Err(ContractError::Vesting {
+                    description: "Vesting must be scheduled from timestamp, not block height"
+                        .to_string(),
+                })
+            }
+            Expiration::Never {} => return Err(ContractError::Vesting {
+                description:
+                    "Vesting must be scheduled from some timestamp, but there wasn't any provided"
+                        .to_string(),
+            }),
+            Expiration::AtTime(timestamp) => timestamp,
+        };
+
+    let transfer_message = Cw20Contract(credits_address.clone())
+        .call(Cw20ExecuteMsg::Transfer {
+            recipient: info.sender.to_string(),
+            amount,
+        })
+        .map_err(ContractError::Std)?;
+    let vesting_message = WasmMsg::Execute {
+        contract_addr: credits_address.to_string(),
+        msg: to_binary(&AddVesting {
+            address: info.sender.to_string(),
+            amount,
+            start_time: vesting_start_time.nanos(),
+            duration: VESTING_DURATION,
+        })?,
+        funds: vec![],
+    };
+    let res = Response::new()
+        .add_message(transfer_message)
+        .add_message(vesting_message)
+        .add_attributes(vec![
+            attr("action", "claim"),
+            attr("stage", stage.to_string()),
+            attr("address", info.sender.to_string()),
+            attr("amount", amount),
+        ]);
+    Ok(res)
+}
+
+pub fn execute_withdraw_all(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+) -> Result<Response, ContractError> {
+    // authorize owner
+    let cfg = CONFIG.load(deps.storage)?;
+    let owner = cfg.owner.ok_or(ContractError::Unauthorized {})?;
+    if info.sender != owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    if !STAGE_PAUSED.load(deps.storage, 1)?
+        && env.block.time
+            <= match STAGE_EXPIRATION.load(deps.storage, 1)? {
+                Expiration::AtTime(timestamp) => timestamp,
+                _ => {
+                    return Err(ContractError::Std(StdError::generic_err(
+                        "withdraw_all only works if AtTime expiration is set",
+                    )))
+                }
+            }
+            .plus_nanos(VESTING_DURATION)
```

Comment:
Why do you add `VESTING_DURATION` here?

Reply:
yeah, there must be a design misconception: users should be capable of claiming tokens up to the end of vesting period, and since then `withdraw_all` is executable. what I can see from the code is:
- `claim` is executable only [until](https://github.com/neutron-org/neutron-tge-contracts/blob/bb61d51b47a1d106b38efa43e8a15654b47a0526/contracts/cw20-merkle-airdrop/src/contract.rs#L256) `STAGE_EXPIRATION` is reached;
- `withdraw_all` is executable starting with `STAGE_EXPIRATION + VESTING_DURATION`.

`VESTING_DURATION` should be taken into account when setting `STAGE_EXPIRATION` to both disable claims and enable withdrawal at the same time, and in this case `plus_nanos(VESTING_DURATION)` is redundant. please correct me if I'm wrong

Reply:
I have fixed all vesting timestamps logic, so it should be good now.